File: lib/domain/cultivation_plan/gateways/optimization_plan_read_gateway.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Gateways
      class OptimizationPlanReadGateway
        PlanCore = Dtos::OptimizationPlanReadPlanCoreSnapshot

        def find_optimization_plan_core_snapshot_by_plan_id(plan_id:)
          raise NotImplementedError, "#{self.class}##{__method__}"
        end

        def find_optimization_weather_location_by_plan_id(plan_id:)
          raise NotImplementedError, "#{self.class}##{__method__}"
        end

        def find_optimization_farm_weather_by_plan_id(plan_id:)
          raise NotImplementedError, "#{self.class}##{__method__}"
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/gateways/optimization_plan_read_gateway.rs
//! Ruby: `Domain::CultivationPlan::Gateways::OptimizationPlanReadGateway`

use crate::cultivation_plan::dtos::OptimizationPlanReadPlanCoreSnapshot;
use crate::weather_data::dtos::{FarmWeatherPrediction, WeatherLocation};

/// Ruby: `Domain::CultivationPlan::Gateways::OptimizationPlanReadGateway`
pub trait OptimizationPlanReadGateway: Send + Sync {
    fn find_optimization_plan_core_snapshot_by_plan_id(
        &self,
        plan_id: i64,
    ) -> Result<OptimizationPlanReadPlanCoreSnapshot, Box<dyn std::error::Error + Send + Sync>>;

    fn find_optimization_weather_location_by_plan_id(
        &self,
        plan_id: i64,
    ) -> Result<Option<WeatherLocation>, Box<dyn std::error::Error + Send + Sync>>;

    fn find_optimization_farm_weather_by_plan_id(
        &self,
        plan_id: i64,
    ) -> Result<Option<FarmWeatherPrediction>, Box<dyn std::error::Error + Send + Sync>>;
}
